### PR TITLE
Support default args

### PR DESCRIPTION
**Pull request by LumaKernel**
Fix https://github.com/vim-jp/vim-vimlparser/issues/154

Is there any spec about Compiler AST ? I broke it a little around `:function`.

MEMO: https://github.com/vim/vim/blob/v8.2.0246/src/userfunc.c#L199-L236

### CODE REVIEW

File: test/test_xxx_funcarg_last_comma.ok [hunk 1]
```diff
@@ -1,2 +1,2 @@
-(function (F a))
-(function (G a))
+(function (F) (a))
+(function (G) (a))
```
Comment (mattn) [inline test/test_xxx_funcarg_last_comma.ok]: I wonder that this is breaking AST for function node?

File: test/test_err_funcarg_default.vim [hunk 2]
```diff
@@ -0,0 +1,2 @@
+function Foo(abc=1, xyz)
+endfunction
```
Comment (mattn) [inline test/test_err_funcarg_default.vim]: Please add test for:

```
function X(foo=1+2)
  return a:foo
endfunction
```

File: test/test_xxx_funcarg_last_comma.ok (hunk 1 above)
Comment (LumaKernel) [reply test/test_xxx_funcarg_last_comma.ok]: Changed.
from
`(function (F) (a))`
to
`(function (F (a)))`

File: test/test_err_funcarg_default.vim (hunk 2 above)
Comment (LumaKernel) [reply test/test_err_funcarg_default.vim]: Added test/test_funcarg_default.vim

File: test/test_xxx_funcarg_last_comma.ok (hunk 1 above)
Comment (LumaKernel) [reply test/test_xxx_funcarg_last_comma.ok]: Update

`function! F(a, b = 2)` : `(function (F (a) (b 2)))`
`function! F(a = 1 + 1, b = 2)` : `(function (F (a (+ 1 1)) (b 2)))`
`function! F(a = 1, b = 2, ...)` : `(function (F (a 1) (b 2) . ...))`

Comment (LumaKernel) [reply test/test_xxx_funcarg_last_comma.ok]: Ah, realized that to omit parenthesis if there's no default args resolves around backward compatibility. And that seems preferable.

Comment (LumaKernel) [reply test/test_xxx_funcarg_last_comma.ok]: Update

`function! F(a, b = 2)` : `(function (F a (b 2)))`

**Comment by mattn** [comment]
Seems good to me. Pass to another look.

File: test/test1.vim [hunk 3]
```diff
@@ -2,6 +2,9 @@
 function s:foo(a, b, ...)
   return 0
 endfunction
+function s:bar(a = 1, b = 2, ...)
+  return 0
+endfunction
```
Comment (tyru) [inline test/test1.vim]: Please add this too.
```vim
function s:baz(a, b = 2, ...)
  return 0
endfunction
```

**Comment by tyru** [comment]
Thanks!

**Pull request merged**